plugin/src/automation/lift/liftrules.ml

```diff
@@ -159,35 +163,78 @@ let is_packed_constr c env sigma trm =
          let constr = constrs.(i - 1) in
          let constr_f = first_fun (unpack (zoom_term zoom_lambda_term env constr)) in
          if equal constr_f f && List.length args = arity constr then
-           sigma, Some (i - 1, args)
+           sigma, Some (i - 1, args, trm)
          else
            sigma, None
       | _ ->
          sigma, None
     else
       sigma, None
   in
-  if isConstruct trm || (isApp trm && l.is_fwd) then
-    is_packed_inductive_constr (fun _ -> true) id trm
-  else
-    match l.orn.kind with
-    | Algebraic _ ->
-       is_packed_inductive_constr (is_packed c) last_arg trm
-    | SwapConstruct _ ->
+  match l.orn.kind with
+  | Algebraic _ ->
+     if isConstruct trm || (isApp trm && l.is_fwd) then
        is_packed_inductive_constr (fun _ -> true) id trm
-    | CurryRecord ->
-       if is_packed c trm then
-          let sigma_right, args_opt = type_is_from c env trm sigma in
-          if Option.has_some args_opt then
-            let sigma = sigma_right in
-            let constr = constrs.(0) in
-            let pms = Option.get args_opt in
-            let args = pair_projections_eta_rec_n trm (arity constr - List.length pms) in
-            sigma, Some (0, List.append pms args)
-          else
-            sigma, None
+     else
+       is_packed_inductive_constr (is_packed c) last_arg trm
+  | SwapConstruct _ ->
+     is_packed_inductive_constr (fun _ -> true) id trm
+  | CurryRecord ->
+      if isConstruct trm || (isApp trm && l.is_fwd) then
+        is_packed_inductive_constr (fun _ -> true) id trm
+      else if is_packed c trm then
+        let sigma_right, args_opt = type_is_from c env trm sigma in
+        if Option.has_some args_opt then
+          let sigma = sigma_right in
+          let constr = constrs.(0) in
+          let pms = Option.get args_opt in
+          let args = pair_projections_eta_rec_n trm (arity constr - List.length pms) in
+          sigma, Some (0, List.append pms args, trm)
+        else
+          sigma, None
+      else
+        sigma, None
+  | UnpackSigma ->
+     (*
+      * TODO
+      * No clue what I'm doing yet! Do not merge this!
```

Comment:
OK now that we know what we're doing, clean up this comment and make it actually explain